sdk/go/pulumi/provider.go

```diff
@@ -0,0 +1,792 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pulumi
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/logging"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/rpcutil"
+
+	pbempty "github.com/golang/protobuf/ptypes/empty"
+	pulumirpc "github.com/pulumi/pulumi/sdk/v2/proto/go"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// ProviderMain is the typical entrypoint for a resource provider plugin. Using it isn't required but can cut down
+// significantly on the amount of boilerplate necessary to fire up a new resource provider.
+func ProviderMain(name, version string, args ProviderArgs) error {
+	if name == "" {
+		return errors.New("name must not be empty")
+	}
+	if version == "" {
+		return errors.New("version must not be empty")
+	}
+
+	var tracing string
+	flag.StringVar(&tracing, "tracing", "", "Emit tracing to a Zipkin-compatible tracing endpoint")
+	flag.Parse()
+
+	// Initialize loggers before going any further.
+	logging.InitLogging(false, 0, false)
+	cmdutil.InitTracing(name, name, tracing)
+
+	// Read the non-flags args and connect to the engine.
+	fargs := flag.Args()
+	if len(fargs) == 0 {
+		return errors.New("fatal: could not connect to host RPC; missing argument")
+	}
+	engineConn, err := grpc.Dial(
+		fargs[0],
+		grpc.WithInsecure(),
+		rpcutil.GrpcChannelOptions(),
+	)
+	if err != nil {
+		return errors.Wrap(err, "connecting to engine over RPC")
+	}
+
+	prov := &providerServer{
+		args:       args,
+		engineConn: engineConn,
+	}
+
+	// Fire up a gRPC server, letting the kernel choose a free port for us.
+	port, done, err := rpcutil.Serve(0, nil, []func(*grpc.Server) error{
+		func(srv *grpc.Server) error {
+			pulumirpc.RegisterResourceProviderServer(srv, prov)
+			return nil
+		},
+	}, nil)
+	if err != nil {
+		return errors.Errorf("fatal: %v", err)
+	}
+
+	// The resource provider protocol requires that we now write out the port we have chosen to listen on.
+	fmt.Printf("%d\n", port)
+
+	// Finally, wait for the server to stop serving.
+	if err := <-done; err != nil {
+		return errors.Errorf("fatal: %v", err)
+	}
+
+	return nil
+}
+
+type providerServer struct {
+	version       string
+	args          ProviderArgs
+	engineConn    *grpc.ClientConn
+	keepSecrets   bool
+	keepResources bool
+}
+
+// ProviderArgs allows specifying implementations of the Provider interface for orchestrating resource create, read,
+// update, and delete operations.
+type ProviderArgs struct {
```

Comment:
I'd originally wanted to name this `Provider`, but we already have a `Provider` identifier in this package (a function for specifying the `Provider` resource option).

Reply:
No longer applicable.